### PR TITLE
Add CI

### PR DESCRIPTION
**Pull request by gkaemmer**
The tests are written in python (much the same as they are currently), and operate by spinning up a server in either the `flask-sqlalchemy` or `rails` folders, restoring fixtures by hitting a `/_reset` endpoint, and then running through test cases using plain old `requests` HTTP methods.

Temporarily basing this PR on `gkaemmer/gitclub-backends`. Figure that we can merge this to main afterwards.

### CODE REVIEW

File: backends/flask-sqlalchemy/app/__init__.py [hunk 1]
```diff
@@ -49,6 +55,19 @@ def handle_forbidden(*_):
     def handle_not_found(*_):
         return {"message": "Not Found"}, 404
 
+    @app.route("/_reset", methods=["POST"])
+    def reset_data():
+        # Called during tests to reset the database
+        session = Session()
+        tables = session.execute(
```
Comment (samscott89) [inline backends/flask-sqlalchemy/app/__init__.py:62]: Would be a little better to do this via SQLAlchemy.

I think you can get all tables with `Base.metadata.tables` and then use `stmt = table.delete()` and `conn.execute(stmt)`.

File: backends/flask-sqlalchemy/app/routes/issues.py [hunk 2]
```diff
@@ -26,8 +26,14 @@ def create(org_id, repo_id):
     issue = Issue(title=payload["title"], repo=repo)
     # check_permission("create", issue)  # TODO(gj): validation check; maybe unnecessary.
     g.session.add(issue)
+    g.session.flush()
+    # For some reason, this method raises a sqlalchemy ObjectDeletedError
+    # if you try to call `issue.repr()` after the commit.
+    # TODO(gkaemmer): figure out why this only happens during tests, and why it
+    # only happens on this endpoint.
+    repr = issue.repr()
```
Comment (samscott89) [inline backends/flask-sqlalchemy/app/routes/issues.py:34]: I think I've seen this before. The problem I saw was that if you access a record after committing it, SQLAlchemy will actually go and grab it from the database (I presume to make sure you're getting whatever columns are automatically set for the model?). But when you do that, we apply authorization. Since there's no permissions specified for `Issue`, you get deny by default for it.

The other routes might handle this.

No idea why this is only during tests.

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+def is_port_open(port):
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+        result = sock.connect_ex(("127.0.0.1", port))
+        return result == 0
+
+
+def ensure_port_5000_is_open(process):
+    sleep(0.5)
+    while not is_port_open(5000):
```
Comment (samscott89) [inline tests/conftest.py:37]: Any chance this could infinitely loop?

File: backends/flask-sqlalchemy/app/__init__.py [hunk 4]
```diff
@@ -1,10 +1,12 @@
 import functools
+import os
 
 from flask import g, Flask, session as flask_session
 from werkzeug.exceptions import BadRequest, Forbidden, NotFound
 
 from sqlalchemy import create_engine
 from sqlalchemy.orm import sessionmaker
+from sqlalchemy.pool import StaticPool
```
Comment (samscott89) [inline backends/flask-sqlalchemy/app/__init__.py:9]: Why is this needed? Could we get a comment explaining why below?

**Review by samscott89 — COMMENTED** [review]
I don't love the approach of resetting the database each time. But I appreciate the simplicity of it.

Is there a chance that concurrent tests would clash? E.g. what if the database was reset after one of the tests created/deleted something?

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        run: black --check backends/flask-sqlalchemy
+      - name: Check Python formatting in tests
+        run: black --check tests
+
+      ## Ruby
+      - name: Install Ruby + gems
+        uses: ruby/setup-ruby@v1
+        with:
+          bundler-cache: true
+          ruby-version: 3.0
```
[the remark this answers is not in the document]
Comment (gkaemmer) [reply .github/workflows/test.yml:32]: Doesn't make much of a difference, I just chose to use the most up-to-date stable version. The app itself doesn't use any ruby 3 features, so I could downgrade to 2.X if we'd prefer.

**Comment by gkaemmer** [comment]
> Is there a chance that concurrent tests would clash? E.g. what if the database was reset after one of the tests created/deleted something?

@samscott89 the tests run in sequence (not in parallel), so this isn't an issue for now.

If we did eventually want to run the tests in parallel, we might have to get fancier. I think doing that would require us either to re-write the test suite so that each test uses data unrelated data (instead of a shared set of fixtures) OR have each test spin up its own server (spinning up a new server would add a lot of overhead, especially for rails which takes ~2-3 seconds to start up)

File: tests/conftest.py (hunk 3 above)
Comment (gkaemmer) [reply tests/conftest.py:37]: Good call, will add a timeout

File: backends/flask-sqlalchemy/app/routes/issues.py (hunk 2 above)
Comment (gkaemmer) [reply backends/flask-sqlalchemy/app/routes/issues.py:34]: Oh yeah confirmed that adding `Issue: "read"` to the session decorator fixes this issue.

Feels kind of weird to me. If I tried to access a totally unrelated model using `g.session` (i.e. `IssueComment` or something), would it also return an empty resultset?

File: .github/workflows/test.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (samscott89) [reply .github/workflows/test.yml:32]: As someone who only installed 2.4 because that's what we require... Yes please :)

File: backends/flask-sqlalchemy/app/__init__.py [hunk 6]
```diff
@@ -49,6 +49,16 @@ def handle_forbidden(*_):
     def handle_not_found(*_):
         return {"message": "Not Found"}, 404
 
+    @app.route("/_reset", methods=["POST"])
+    def reset_data():
+        # Called during tests to reset the database
+        session = Session()
+        Base.metadata.drop_all(bind=engine)
```
Comment (samscott89) [inline backends/flask-sqlalchemy/app/__init__.py:56]: Oh lol even better!

Comment (gkaemmer) [reply backends/flask-sqlalchemy/app/__init__.py:56]: 💣

File: tests/conftest.py (hunk 3 above)
Comment (samscott89) [reply tests/conftest.py:39]: Thanks!

**Review by samscott89 — APPROVED** [review]
LGTM! Good work.
